### ISSUE TITLE
Identically named type annotation on attrs field is resolved to the field name

### ISSUE THREAD

**Issue by injust** [astral-sh/ty#3175]
### Summary

```python
import uuid

from attrs import define, field


@define
class Foo:
    uuid: uuid.UUID = field()  # Object of type `dataclasses.Field` has no attribute `UUID` (ty unresolved-attribute)

    def bar(self, thing: uuid.UUID) -> None:  # Object of type `dataclasses.Field` has no attribute `UUID` (ty unresolved-attribute)
        pass
```

The errors don't happen if you remove `= field()`. The 2nd error disappears if you move `bar` into a child class.

### Version

ty v0.0.26 (940305127 2026-03-26)

**Comment by sharkdp** [comment]
Thank you for reporting this.

I can not reproduce this (I tried adding a `Foo(uuid.uuid1()).uuid` line to check if the field is available). Can you show the full file and add some more details about your setup?

**Comment by injust** [comment]
I was able to reproduce using that snippet in a new Python 3.14.3 venv, with no special configuration (the repo was initialized with `uv init`).

The only thing I installed was attrs 26.1.0.

**Comment by injust** [comment]
> I tried adding a `Foo(uuid.uuid1()).uuid` line to check if the field is available

I tested this in the REPL and works as expected:

```
>>> Foo(uuid.uuid1())
Foo(uuid=UUID('1f950b2e-2c0c-11f1-8cdf-d7dc570d4f67'))
>>> Foo(uuid.uuid1()).uuid
UUID('209a1e1a-2c0c-11f1-8cdf-d7dc570d4f67')
```

So I believe ty is resolving the type annotation incorrectly.

**Comment by sharkdp** [comment]
Ok, after your edit, I can now also reproduce this. The problem is that you overwrite the symbol `uuid` in the class scope of `Foo`. So in the type annotation `uuid.UUID` for the parameter `thing`, `uuid` refers to the field, not the module. This is not a problem at runtime (in 3.14), since the annotation is not evaluated.

This is not a problem in ty, this is a problem in your code.

**Comment by injust** [comment]
attrs isn't necessary for this. Also happens with dataclasses:

```python
import uuid
from dataclasses import dataclass, field


@dataclass
class Foo:
    uuid: uuid.UUID = field()  # Object of type `dataclasses.Field` has no attribute `UUID` (ty unresolved-attribute)

    def bar(self, thing: uuid.UUID) -> None:  # Object of type `dataclasses.Field` has no attribute `UUID` (ty unresolved-attribute)
        pass
```

**Comment by injust** [comment]
> This is not a problem in ty, this is a problem in your code.

I'm not sure, then. basedpyright infers the correct types.

https://basedpyright.com/?typeCheckingMode=all&code=JYWwDg9gTgLgBAVwcAJgKAGZQiOKCGM%2BAxgDb4DOFAphXKJLHoSeVQDRwbDWnppoAAgSJlKFNGKpwAYhAgAuNHBWJkKBWtQA6AKq6AkgBE4AXi48%2BACgCUA1XmoY4AI3xQrNUhk4wAFsAAdgDmmkg6%2BsY2cAC0AHxwAHIQgdRKDg5g4mhAA

**Comment by injust** [comment]
Same thing happens in Python 3.13 with `from __future__ import annotations`. basedpyright infers the correct types, but not ty.

**Comment by sharkdp** [comment]
Please see https://github.com/astral-sh/ty/issues/900 and https://github.com/astral-sh/ty/issues/2430 for related issues. https://github.com/astral-sh/ty/issues/900#issuecomment-3151616356 describes our reasoning and our general advise is to avoid shadowing names: https://github.com/astral-sh/ty/issues/2430#issuecomment-3731059444

**Comment by injust** [comment]
If I'm reading https://github.com/astral-sh/ty/issues/900#issuecomment-3151616356 correctly, ty should match the runtime behaviour of annotations in 3.14?

```python
>>> annotationlib.get_annotations(Foo)
{'uuid': <class 'uuid.UUID'>}
>>> annotationlib.get_annotations(Foo.bar)
{'thing': <class 'uuid.UUID'>, 'return': None}
```

**Comment by sharkdp** [comment]
> If I'm reading [#900 (comment)](https://github.com/astral-sh/ty/issues/900#issuecomment-3151616356) correctly, ty should match the runtime behaviour of annotations in 3.14

The linked comment explicitly states that we *"don't consider `typing.get_type_hints` or any other pre-3.14 implementation of "resolve stringified annotations back to runtime values" to be a useful guide to how stringified annotations should resolve in edge cases"*.

**Comment by injust** [comment]
Well I am only referring to 3.14 here. That paragraph also continues with:

> That means the only runtime behavior I consider to be a useful guide to deferred name resolution is the behavior of 3.14 without `from __future__ import annotations`.

From what I read, the advice to avoid shadowing names was due to edge cases with stringified annotations pre-3.14. In 3.14, it sounds like ty should be matching runtime annotations.

**Comment by sharkdp** [comment]
I think the runtime behavior is simply not a good guide here. It's not even consistent. If you run this variant of your example on 3.14:
```py
import annotationlib
import uuid
from dataclasses import dataclass, field


@dataclass
class Foo:
    uuid: uuid.UUID = field()

    def bar(self, thing: uuid.UUID) -> None:
        pass

    print(annotationlib.get_annotations(bar))
```
you will get an error that reflects what ty reports:
```
AttributeError: 'Field' object has no attribute 'UUID'
```

mypy and pyrefly also emit errors on your original example. I would also argue that it's highly irritating for human readers of the code.

**Comment by svioletg** [comment]
Was searching to see if this issue had already been raised, adding on that this doesn't need to be a dataclass or use @ property at all, this'll happen with any method:

```py
from typing import Any

class A:
    def __init__(self, x: int) -> None:
        self.x = x

    def dict(self) -> dict[str, Any]:
        return {'x': self.x}

a = A(1)
print(a.dict())
```

This example runs without error (printing `{'x': 1}`), passes mypy, pyrefly, and basedpyright's checks (basedpyright does complain about using Any, but that's unrelated), but does *not* pass ty's check, resulting in this error:

```
error[invalid-type-form]: Invalid subscript of object of type `def dict(self) -> Unknown` in a return type annotation
 --> mre.py:7:23
  |
7 |     def dict(self) -> dict[str, Any]:
  |                       ^^^^^^^^^^^^^^
```

I experienced a more practical version of this issue defining a `round()` method on my coordinate class, then getting type errors when trying to use the built-in `round` as the default value for another method's parameter.

**Comment by sharkdp** [comment]
> I experienced a more practical version of this issue defining a `round()` method on my coordinate class, then getting type errors when trying to use the built-in `round` as the default value for another method's parameter.

Hm, a default value is not in type expression position, like the other examples here. Can you show a minified example of this?

**Comment by carljm** [comment]
> This example runs without error (printing `{'x': 1}`), passes mypy, pyrefly, and basedpyright's checks (basedpyright does complain about using Any, but that's unrelated), but does _not_ pass ty's check

This is #1747. ty is correct here and the other type checkers are wrong, and there is currently an open proposal on the typing spec / conformance suite to specify that all type checkers should have ty's behavior in this situation.

If you add the line `print(A.dict.__annotations__)` to the end of your example, you will see a runtime exception that exactly mirrors ty's diagnostic:

```
Traceback (most recent call last):
  File "<string>", line 22, in <module>
    runpy.run_path(entrypoint, run_name="__main__")
    ~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<frozen runpy>", line 287, in run_path
  File "<frozen runpy>", line 98, in _run_module_code
  File "<frozen runpy>", line 88, in _run_code
  File "main.py", line 12, in <module>
    print(A.dict.__annotations__)
          ^^^^^^^^^^^^^^^^^^^^^^
  File "main.py", line 7, in __annotate__
    def dict(self) -> dict[str, Any]:
                      ~~~~^^^^^^^^^^
TypeError: 'function' object is not subscriptable
```

 So ty is alerting you to a real runtime issue that all other type checkers are missing, it's just that the issue only surfaces at runtime if you actually introspect the return annotation.

**Comment by injust** [comment]
> I think the runtime behavior is simply not a good guide here. It's not even consistent.

Fair enough.
